Firestore/src/BulkWriter.php

```diff
@@ -0,0 +1,1257 @@
+    use ValidateTrait;
+
+    const TYPE_UPDATE = 'update';
+    const TYPE_SET = 'set';
+    const TYPE_CREATE = 'create';
+    const TYPE_DELETE = 'delete';
+    const TYPE_TRANSFORM = 'transform';
+
+    /** The maximum number of writes that can be in a single batch. */
+    const MAX_BATCH_SIZE = 20;
+
+    /** The maximum number of writes that can be in a batch containing retries. */
+    const RETRY_MAX_BATCH_SIZE = 20;
+
+    /**
+     * The maximum number of retries that will be attempted with backoff before stopping all retry
+     * attempts.
+     */
+    const MAX_RETRY_ATTEMPTS = 10;
+
+    /**
+     * The default initial backoff time in milliseconds after an error. Set to 1s according to
+     * https://cloud.google.com/apis/design/errors.
+     */
+    const DEFAULT_BACKOFF_INITIAL_DELAY_MS = 1000;
+
+    /** The default maximum backoff time in milliseconds when retrying an operation. */
+    const DEFAULT_BACKOFF_MAX_DELAY_MS = 60 * 1000;
+
+    /** The default factor to increase the backup by after each failed attempt. */
+    const DEFAULT_BACKOFF_FACTOR = 1.5;
+
+    /**
+     * The default jitter to apply to the exponential backoff used in retries.
+     * For example, a factor of 0.3 means a 30% jitter is applied.
+     */
+    const DEFAULT_JITTER_FACTOR = 0.3;
+
+    /**
+     * @var ConnectionInterface
+     */
+    private $connection;
+
+    /**
+     * @var ValueMapper
+     */
+    private $valueMapper;
+
+    /**
+     * @var string
+     */
+    private $database;
+
+    /**
+     * @var string|null
+     */
+    private $transaction;
+
+    /**
+     * @var bool
+     */
+    private $isLegacyWriteBatch;
+
+    /**
+     * @var int
+     */
+    private $maxBatchSize;
+
+    /**
+     * @var array
+     */
+    private $writes = [];
+
+    /**
+     * @var array
+     */
+    private $finalResponse = [];
+
+    /**
+     * @var array Failed rescheduled mutations.
+     * Each object to have these fields:
+     * 'num_failed_attempts': int
+     * 'scheduled_in_millis': int
+     * 'backoff_in_millis': int
+     */
+    private $retryScheduledWrites = [];
+
+    /**
+     * @var array All unique documents added to mutate.
+     */
+    private $unique_documents = [];
+
+    /**
+     * @var bool Whether this BulkWriter instance is closed.
+     * Once closed, it cannot be opened again.
+     */
+    private $closed;
+
+    /**
+     * @var bool Whether BulkWriter greedily sends operations via
+     * [https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1beta1#batchwriterequest](BatchWriteRequest)
+     * when sufficient number of operations are enqueued.
+     */
+    private $greedilySend;
+
+    /**
+     * @param ConnectionInterface $connection A connection to Cloud Firestore
+     * @param ValueMapper $valueMapper A Value Mapper instance
+     * @param string $database The current database
+     * @param array|string|null $options [optional] Configuration options as array. {
```

Comment:
I wanted to test this out, and unfortunately it seems as though our documentation generator isn't equipped to handle an array + other type param block, e.g.,:

<img width="1017" alt="Screen Shot 2022-09-28 at 9 37 38 AM" src="https://user-images.githubusercontent.com/2079879/192838240-7c45ad4b-e34f-4b4d-8c88-5e3afc364282.png">

I'm thinking through options/solutions.

Reply:
@dwsupplee it was `string|null` in the older WriteBatch, [ref link](https://github.com/googleapis/google-cloud-php/blob/main/Firestore/src/WriteBatch.php#L88). 

Also, I see we have array|**_SomeClass_** in gapic. Will it be better to put into a new class?

Reply:
Understood, TODO: reformat according to our docs. ref: https://googleapis.github.io/google-cloud-php/#/docs/cloud-firestore/main/firestore/writebatch?method=__construct

Reply:
Also add this class in `/docs/contents` so that `/docs/json` is autogenerated.

Reply:
Hi @dwsupplee , now the bulkwriter docs looks like this:
<img width="1114" alt="Screenshot 2022-10-20 at 17 19 57" src="https://user-images.githubusercontent.com/7369612/196941341-ea706086-1bbc-4c58-89cb-3f9db7cef9be.png">